File: ocaml/src/p0055.ml
(* Project Euler: Problem 55 *)

open Core

let is_rychrel num =
  let cnt = 50 in
  let rev_bignum n = Z.(to_string n |> String.rev |> of_string) in
  let is_palindromic n = Z.equal n (rev_bignum n) in
  Sequence.(
    unfold ~init:(Z.of_int num) ~f:(fun n ->
      match Z.(n + rev_bignum n) with
      | next_n when is_palindromic next_n -> None
      | next_n -> Some (next_n, next_n))
    |> Fun.flip take cnt
    |> length)
  |> ( = ) cnt
;;

let compute upper = Sequence.(range 1 upper ~stop:`inclusive |> count ~f:is_rychrel)
let solve () = compute 10_000 |> Int.to_string

(* Test *)

let%test_unit "10_000" = [%test_eq: int] (compute 10_000) 249

File: rust/src/bin/p0055.rs
// Project Euler: Problem 55

use num_bigint::BigUint;

euler::run_solver!(55);

fn solve() -> String {
    compute(10_000).to_string()
}

fn compute(upper: u32) -> usize {
    debug_assert!(upper > 0);

    (1..=upper)
        .filter(|&n| is_rychrel(BigUint::from(n)))
        .count()
}

fn is_rychrel(mut n: BigUint) -> bool {
    let mut tmp = rev_biguint(&n);
    for _ in 0..50 {
        n += tmp;
        tmp = rev_biguint(&n);
        if n == tmp {
            return false;
        }
    }

    true
}

fn rev_biguint(n: &BigUint) -> BigUint {
    let rev_string = |s: &str| s.chars().rev().collect::<String>();

    BigUint::parse_bytes(rev_string(&BigUint::to_str_radix(n, 10)).as_bytes(), 10).unwrap()
}

#[cfg(test)]
mod tests {
    use super::compute;

    #[test]
    fn p0055_upper_10000() {
        assert_eq!(compute(10_000), 249);
    }
}
